File: backend/services/omniforge_polyglot_registry.py
"""Polyglot plugin registry — languages, domains, build/run profiles for OmniForge Engine."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal

DomainId = Literal[
    "web_saas",
    "web_ecommerce",
    "web_portfolio",
    "mobile_android",
    "mobile_ios",
    "mobile_flutter",
    "desktop_win",
    "desktop_mac",
    "desktop_linux",
    "game_2d",
    "game_3d",
    "microservice",
    "data_science",
    "browser_extension",
    "enterprise",
]

LanguageId = Literal[
    "html", "css", "javascript", "typescript", "python", "c", "cpp", "csharp",
    "java", "kotlin", "swift", "dart", "php", "go", "rust", "ruby", "lua", "r", "sql", "bash", "powershell",
]


@dataclass(frozen=True)
class LanguagePlugin:
    id: LanguageId
    label: str
    extensions: tuple[str, ...]
    monaco_id: str
    compile_cmd: str | None = None
    run_cmd: str | None = None


@dataclass(frozen=True)
class DomainProfile:
    id: DomainId
    label: str
    default_languages: tuple[LanguageId, ...]
    scaffold_adapter: str
    preview_mode: str = "web_blob"


LANGUAGE_PLUGINS: dict[str, LanguagePlugin] = {
    "html": LanguagePlugin("html", "HTML", (".html", ".htm"), "html"),
    "css": LanguagePlugin("css", "CSS", (".css", ".scss"), "css"),
    "javascript": LanguagePlugin("javascript", "JavaScript", (".js", ".mjs", ".cjs"), "javascript", run_cmd="node {file}"),
    "typescript": LanguagePlugin("typescript", "TypeScript", (".ts", ".tsx"), "typescript", compile_cmd="tsc", run_cmd="node {file}"),
    "python": LanguagePlugin("python", "Python", (".py",), "python", run_cmd="python {file}"),
    "c": LanguagePlugin("c", "C", (".c", ".h"), "c", compile_cmd="gcc {file} -o {out}"),
    "cpp": LanguagePlugin("cpp", "C++", (".cpp", ".hpp", ".cc"), "cpp", compile_cmd="g++ {file} -o {out}"),
    "csharp": LanguagePlugin("csharp", "C#", (".cs",), "csharp", compile_cmd="dotnet build"),
    "java": LanguagePlugin("java", "Java", (".java",), "java", compile_cmd="javac {file}", run_cmd="java {class}"),
    "kotlin": LanguagePlugin("kotlin", "Kotlin", (".kt", ".kts"), "kotlin"),
    "swift": LanguagePlugin("swift", "Swift", (".swift",), "swift"),
    "dart": LanguagePlugin("dart", "Dart", (".dart",), "dart", run_cmd="dart run {file}"),
    "php": LanguagePlugin("php", "PHP", (".php",), "php", run_cmd="php {file}"),
    "go": LanguagePlugin("go", "Go", (".go",), "go", run_cmd="go run {file}"),
    "rust": LanguagePlugin("rust", "Rust", (".rs",), "rust", compile_cmd="cargo build"),
    "ruby": LanguagePlugin("ruby", "Ruby", (".rb",), "ruby", run_cmd="ruby {file}"),
    "lua": LanguagePlugin("lua", "Lua", (".lua",), "lua"),
    "r": LanguagePlugin("r", "R", (".r", ".R"), "r"),
    "sql": LanguagePlugin("sql", "SQL", (".sql",), "sql"),
    "bash": LanguagePlugin("bash", "Bash", (".sh",), "shell"),
    "powershell": LanguagePlugin("powershell", "PowerShell", (".ps1",), "powershell"),
}

DOMAIN_PROFILES: dict[str, DomainProfile] = {
    "web_saas": DomainProfile("web_saas", "Web SaaS", ("typescript", "javascript", "html", "css", "python"), "app-builder", "web_blob"),
    "web_ecommerce": DomainProfile("web_ecommerce", "E-Commerce", ("typescript", "html", "css", "python"), "business-site-maker", "web_blob"),
    "web_portfolio": DomainProfile("web_portfolio", "Portfolio", ("html", "css", "javascript"), "app-builder", "web_blob"),
    "mobile_flutter": DomainProfile("mobile_flutter", "Flutter Mobile", ("dart",), "app-builder", "device_frame"),
    "mobile_android": DomainProfile("mobile_android", "Android", ("kotlin",), "app-builder", "device_frame"),
    "mobile_ios": DomainProfile("mobile_ios", "iOS", ("swift",), "app-builder", "device_frame"),
    "desktop_win": DomainProfile("desktop_win", "Windows Desktop", ("csharp",), "app-builder", "desktop_frame"),
    "game_2d": DomainProfile("game_2d", "2D Game", ("csharp", "javascript"), "game-dev", "game_preview"),
    "game_3d": DomainProfile("game_3d", "3D Game", ("csharp", "cpp"), "game-dev", "game_preview"),
    "microservice": DomainProfile("microservice", "Microservices", ("go", "python", "typescript"), "app-builder", "api_panel"),
    "data_science": DomainProfile("data_science", "AI / Data Science", ("python", "r", "sql"), "app-builder", "notebook"),
    "browser_extension": DomainProfile("browser_extension", "Browser Extension", ("javascript", "typescript", "html"), "app-builder", "web_blob"),
    "enterprise": DomainProfile("enterprise", "Enterprise", ("java", "csharp", "sql"), "business-site-maker", "web_blob"),
}


def resolve_language(path: str) -> LanguageId | None:
    low = path.lower()
    for plugin in LANGUAGE_PLUGINS.values():
        if any(low.endswith(ext) for ext in plugin.extensions):
            return plugin.id
    return None


def register_language(plugin: LanguagePlugin) -> None:
    """Dynamic plugin registration hook."""
    LANGUAGE_PLUGINS[plugin.id] = plugin


def register_domain(profile: DomainProfile) -> None:
    DOMAIN_PROFILES[profile.id] = profile


def registry_snapshot() -> dict[str, Any]:
    return {
        "languages": [p.id for p in LANGUAGE_PLUGINS.values()],
        "domains": [d.id for d in DOMAIN_PROFILES.values()],
        "language_count": len(LANGUAGE_PLUGINS),
        "domain_count": len(DOMAIN_PROFILES),
    }

File: frontend/lib/omniforge-polyglot-registry.ts
/** Polyglot plugin registry — languages, domains, build profiles (mirrors backend). */

export type LanguageId =
  | "html" | "css" | "javascript" | "typescript" | "python" | "c" | "cpp" | "csharp"
  | "java" | "kotlin" | "swift" | "dart" | "php" | "go" | "rust" | "ruby" | "lua" | "r" | "sql" | "bash" | "powershell";

export type DomainId =
  | "web_saas" | "web_ecommerce" | "web_portfolio" | "mobile_android" | "mobile_ios" | "mobile_flutter"
  | "desktop_win" | "desktop_mac" | "desktop_linux" | "game_2d" | "game_3d" | "microservice"
  | "data_science" | "browser_extension" | "enterprise";

export type LanguagePlugin = {
  id: LanguageId;
  label: string;
  extensions: string[];
  monacoId: string;
  compileCmd?: string;
  runCmd?: string;
};

export type DomainProfile = {
  id: DomainId;
  label: string;
  defaultLanguages: LanguageId[];
  scaffoldAdapter: string;
  previewMode: string;
};

export const LANGUAGE_PLUGINS: Record<LanguageId, LanguagePlugin> = {
  html: { id: "html", label: "HTML", extensions: [".html", ".htm"], monacoId: "html" },
  css: { id: "css", label: "CSS", extensions: [".css", ".scss"], monacoId: "css" },
  javascript: { id: "javascript", label: "JavaScript", extensions: [".js", ".mjs", ".cjs"], monacoId: "javascript", runCmd: "node {file}" },
  typescript: { id: "typescript", label: "TypeScript", extensions: [".ts", ".tsx"], monacoId: "typescript", compileCmd: "tsc", runCmd: "node {file}" },
  python: { id: "python", label: "Python", extensions: [".py"], monacoId: "python", runCmd: "python {file}" },
  c: { id: "c", label: "C", extensions: [".c", ".h"], monacoId: "c" },
  cpp: { id: "cpp", label: "C++", extensions: [".cpp", ".hpp", ".cc"], monacoId: "cpp" },
  csharp: { id: "csharp", label: "C#", extensions: [".cs"], monacoId: "csharp", compileCmd: "dotnet build" },
  java: { id: "java", label: "Java", extensions: [".java"], monacoId: "java" },
  kotlin: { id: "kotlin", label: "Kotlin", extensions: [".kt", ".kts"], monacoId: "kotlin" },
  swift: { id: "swift", label: "Swift", extensions: [".swift"], monacoId: "swift" },
  dart: { id: "dart", label: "Dart", extensions: [".dart"], monacoId: "dart", runCmd: "dart run {file}" },
  php: { id: "php", label: "PHP", extensions: [".php"], monacoId: "php", runCmd: "php {file}" },
  go: { id: "go", label: "Go", extensions: [".go"], monacoId: "go", runCmd: "go run {file}" },
  rust: { id: "rust", label: "Rust", extensions: [".rs"], monacoId: "rust", compileCmd: "cargo build" },
  ruby: { id: "ruby", label: "Ruby", extensions: [".rb"], monacoId: "ruby", runCmd: "ruby {file}" },
  lua: { id: "lua", label: "Lua", extensions: [".lua"], monacoId: "lua" },
  r: { id: "r", label: "R", extensions: [".r", ".R"], monacoId: "r" },
  sql: { id: "sql", label: "SQL", extensions: [".sql"], monacoId: "sql" },
  bash: { id: "bash", label: "Bash", extensions: [".sh"], monacoId: "shell" },
  powershell: { id: "powershell", label: "PowerShell", extensions: [".ps1"], monacoId: "powershell" },
};

export const DOMAIN_PROFILES: Record<DomainId, DomainProfile> = {
  web_saas: { id: "web_saas", label: "Web SaaS", defaultLanguages: ["typescript", "javascript", "html", "css", "python"], scaffoldAdapter: "app-builder", previewMode: "web_blob" },
  web_ecommerce: { id: "web_ecommerce", label: "E-Commerce", defaultLanguages: ["typescript", "html", "css", "python"], scaffoldAdapter: "business-site-maker", previewMode: "web_blob" },
  web_portfolio: { id: "web_portfolio", label: "Portfolio", defaultLanguages: ["html", "css", "javascript"], scaffoldAdapter: "app-builder", previewMode: "web_blob" },
  mobile_flutter: { id: "mobile_flutter", label: "Flutter Mobile", defaultLanguages: ["dart"], scaffoldAdapter: "app-builder", previewMode: "device_frame" },
  mobile_android: { id: "mobile_android", label: "Android", defaultLanguages: ["kotlin"], scaffoldAdapter: "app-builder", previewMode: "device_frame" },
  mobile_ios: { id: "mobile_ios", label: "iOS", defaultLanguages: ["swift"], scaffoldAdapter: "app-builder", previewMode: "device_frame" },
  desktop_win: { id: "desktop_win", label: "Windows Desktop", defaultLanguages: ["csharp"], scaffoldAdapter: "app-builder", previewMode: "desktop_frame" },
  desktop_mac: { id: "desktop_mac", label: "macOS Desktop", defaultLanguages: ["swift"], scaffoldAdapter: "app-builder", previewMode: "desktop_frame" },
  desktop_linux: { id: "desktop_linux", label: "Linux Desktop", defaultLanguages: ["cpp"], scaffoldAdapter: "app-builder", previewMode: "desktop_frame" },
  game_2d: { id: "game_2d", label: "2D Game", defaultLanguages: ["csharp", "javascript"], scaffoldAdapter: "game-dev", previewMode: "game_preview" },
  game_3d: { id: "game_3d", label: "3D Game", defaultLanguages: ["csharp", "cpp"], scaffoldAdapter: "game-dev", previewMode: "game_preview" },
  microservice: { id: "microservice", label: "Microservices", defaultLanguages: ["go", "python", "typescript"], scaffoldAdapter: "app-builder", previewMode: "api_panel" },
  data_science: { id: "data_science", label: "AI / Data Science", defaultLanguages: ["python", "r", "sql"], scaffoldAdapter: "app-builder", previewMode: "notebook" },
  browser_extension: { id: "browser_extension", label: "Browser Extension", defaultLanguages: ["javascript", "typescript", "html"], scaffoldAdapter: "app-builder", previewMode: "web_blob" },
  enterprise: { id: "enterprise", label: "Enterprise", defaultLanguages: ["java", "csharp", "sql"], scaffoldAdapter: "business-site-maker", previewMode: "web_blob" },
};

const pluginByExt = new Map<string, LanguagePlugin>();
for (const plugin of Object.values(LANGUAGE_PLUGINS)) {
  for (const ext of plugin.extensions) pluginByExt.set(ext, plugin);
}

export function resolveLanguagePlugin(path: string): LanguagePlugin | null {
  const low = path.toLowerCase();
  for (const [ext, plugin] of pluginByExt) {
    if (low.endsWith(ext)) return plugin;
  }
  return null;
}

export function registerLanguagePlugin(plugin: LanguagePlugin): void {
  LANGUAGE_PLUGINS[plugin.id] = plugin;
  for (const ext of plugin.extensions) pluginByExt.set(ext, plugin);
}

export function registerDomainProfile(profile: DomainProfile): void {
  DOMAIN_PROFILES[profile.id] = profile;
}
